tiatoolbox/models/architecture/mapde.py

```diff
@@ -0,0 +1,293 @@
+                    0.36754447,
+                    0.5527864,
+                    0.71715724,
+                    0.8,
+                    0.71715724,
+                    0.5527864,
+                    0.36754447,
+                    0.17537889,
+                    0.0,
+                ],
+                [0.0, 0.2, 0.4, 0.6, 0.8, 1.0, 0.8, 0.6, 0.4, 0.2, 0.0],
+                [
+                    0.0,
+                    0.17537889,
+                    0.36754447,
+                    0.5527864,
+                    0.71715724,
+                    0.8,
+                    0.71715724,
+                    0.5527864,
+                    0.36754447,
+                    0.17537889,
+                    0.0,
+                ],
+                [
+                    0.0,
+                    0.1055728,
+                    0.27888975,
+                    0.43431458,
+                    0.5527864,
+                    0.6,
+                    0.5527864,
+                    0.43431458,
+                    0.27888975,
+                    0.1055728,
+                    0.0,
+                ],
+                [
+                    0.0,
+                    0.0,
+                    0.1514719,
+                    0.27888975,
+                    0.36754447,
+                    0.4,
+                    0.36754447,
+                    0.27888975,
+                    0.1514719,
+                    0.0,
+                    0.0,
+                ],
+                [
+                    0.0,
+                    0.0,
+                    0.0,
+                    0.1055728,
+                    0.17537889,
+                    0.2,
+                    0.17537889,
+                    0.1055728,
+                    0.0,
+                    0.0,
+                    0.0,
+                ],
+                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+            ],
+            dtype=np.float32,
+        )
+
+        dist_filter = np.expand_dims(dist_filter, axis=(0, 1))  # NCHW
+        dist_filter = np.repeat(dist_filter, repeats=num_class * 2, axis=1)
+
+        self.min_distance = min_distance
+        self.threshold_abs = threshold_abs
+        self.register_buffer(
+            "dist_filter",
+            torch.from_numpy(dist_filter.astype(np.float32)),
+        )
+        self.dist_filter.requires_grad = False
+
+    def forward(self, input_tensor: torch.Tensor) -> torch.Tensor:
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+        logits, _, _, _ = super().forward(input_tensor)
```

Comment:
I see no `preproc` function used here nor in `micronet` implementation. Have you checked the sanity of the outputs that this architecture generates? Because, as far as I remember, you used image normalization in the input of the model also see you have `preproc` method defined in the `micronet` class but never have used these. 

I would suggest you move the required preprocessing of your model into the `forward` method and name the preprocessing method after unet convention to `_transform` [(see here)](https://github.com/TissueImageAnalytics/tiatoolbox/blob/fd957affe0a6cb08fce3eda1431e5812dacb065a/tiatoolbox/models/architecture/unet.py#L324)

Reply:
Now I can see from the tests that you call preprocessing separately before calling model forward, however, the convention with pytorch model is to include the preprocessing in the forward method.

Reply:
Actually there should be pre-proc and that includes subtracting mean and std.